activerecord/lib/active_record/normalization.rb

```diff
@@ -0,0 +1,151 @@
+    end
+
+    # Normalizes a specified attribute using its declared normalizations.
+    #
+    # ==== Examples
+    #
+    #   class User < ActiveRecord::Base
+    #     normalizes :email, with: -> email { email.strip.downcase }
+    #   end
+    #
+    #   legacy_user = User.find(1)
+    #   legacy_user.email # => " CRUISE-CONTROL@EXAMPLE.COM\n"
+    #   legacy_user.normalize_attribute(:email)
+    #   legacy_user.email # => "cruise-control@example.com"
+    #   legacy_user.save
+    def normalize_attribute(name)
+      # Set the "changed" flag to trigger a type cast.
+      self[name] = self[name]
+    end
+
+    module ClassMethods
+      # Declares a normalization to apply to one or more attributes.
+      # Normalizations are applied when the attributes are assigned or updated,
+      # and the normalized values will be persisted to the database.
+      # Normalizations are also applied to matching keyword arguments of finder
+      # methods. This allows a record to be created and later queried using an
+      # unnormalized value.
+      #
+      # However, to prevent confusion, normalizations are not applied when
+      # attributes are fetched from the database. This means that if a record
+      # was persisted before a normalization was declared, the relevant
+      # attributes will not be normalized until either they are assigned new
+      # values, or they are explicitly migrated via +normalize_attribute+.
+      #
+      # Because normalizations may be applied multiple times, they should be
+      # _idempotent_. In other words, applying a normalization more than once
+      # should have the same result as applying it only once.
+      #
+      # Normalizations, by default, are not applied to +nil+ values. This
+      # behavior can be changed with the +:including_nil_values+ option.
+      #
+      # ==== Options
+      #
+      # * <tt>:with</tt> - The normalization to apply. May be specified as
+      #   either a callable or a Symbol which will be converted to a Proc.
+      # * <tt>:including_nil_values</tt> - Whether to apply normalizations to
+      #   +nil+ values. Defaults to +false+.
+      #
+      # ==== Examples
+      #
+      #   class User < ActiveRecord::Base
+      #     normalizes :email, with: -> email { email.strip.downcase }
+      #   end
+      #
+      #   user = User.create(email: " CRUISE-CONTROL@EXAMPLE.COM\n")
+      #   user.email                  # => "cruise-control@example.com"
+      #
+      #   user = User.find_by(email: "\tCRUISE-CONTROL@EXAMPLE.COM ")
+      #   user.email                  # => "cruise-control@example.com"
+      #   user.email_before_type_cast # => "cruise-control@example.com"
+      #
+      #   User.where(email: "\tCRUISE-CONTROL@EXAMPLE.COM ").first       # => user
+      #   User.where("email = ?", "\tCRUISE-CONTROL@EXAMPLE.COM ").first # => nil
+      def normalizes(*names, including_nil_values: false, with:)
```

Comment:
What do you think about naming `including_nil_values` as `normalize_nil`, as you did on `NormalizedValueType`? Also, thoughts on moving the optional argument to the end of the signature?


```suggestion
      def normalizes(*names, with:, normalize_nil: false)
```

Reply:
Or maybe `allow_nil` as used in `validates`?
```suggestion
      def normalizes(*names, with:, allow_nil: false)
```

Reply:
My intention here was for a `normalizes` declaration to read like natural language.  For example:

```ruby
normalizes :email, including_nil_values: true, with: -> email { email.strip.downcase }
```

Would read as "normalizes email addresses, including nil addresses, with `strip` and `downcase`."

Since `normalizes` doesn't *disallow* values, I'm not sure if `allow_nil` would be appropriate.  If we need something shorter, my preference would be `including_nil`.  (Though I'm also not sure how much there is to gain by making it shorter, since I think it would typically be left to default.)